engine/chassis.ts

```diff
@@ -398,13 +401,14 @@ export class Program {
             samplerTextures,
         ));
 
+        this._inputState = new InputState(shaderAttributes);
         this._layout = TestBase.device.createPipelineLayout(new PipelineLayoutInfo(descriptorSetLayouts));
         this._pipelineStateMap = new Map<number, PipelineState>();
         this._currentPipelineState = null!;
         this.setPipelineState(info.defaultStates || {});
     }
 
-    private _bufferStreamIds (attributes: IShaderAttribute[] = []): number[] {
+    public static bufferStreamIds (attributes: IShaderAttribute[] | Attribute[] = []): number[] {
```

Comment:
The `Attribute` type is a strict sub-type of `IShaderAttribute` so I think accepting `IShaderAttribute[]` as the parameter type would suffice.

https://en.wikipedia.org/wiki/Duck_typing

Reply:
Besides, this function is only used inside this file so it doesn't need to be exposed as public, just declare this as an independent util function outside the class definition will do.